diff -Bru falcon_rel_5_90_188_59_org/src/usbdev/usbdl/Makefile falcon_rel_5_90_188_59/src/usbdev/usbdl/Makefile
--- falcon_rel_5_90_188_59_org/src/usbdev/usbdl/Makefile	2012-10-04 15:41:05.000000000 -0400
+++ falcon_rel_5_90_188_59/src/usbdev/usbdl/Makefile	2012-10-10 17:36:41.471312486 -0400
@@ -46,20 +46,18 @@
 ifeq ($(TARGETENV), linuxmips_be) 
 	OBJECTS += $(OBJDIR)usb_linux.o
 	COMPILE = mips-linux-gcc 
-	LIBS := -L$(LIBUSB_PATH)/.libs $(LIBUSB_PATH)/.libs/libusb.a
-	CFLAGS += -I$(LIBUSB_PATH) -DIL_BIGENDIAN
+	LIBS := -L$(LIBUSB_PATH)/lib $(LIBUSB_PATH)/lib/libusb.a
+	CFLAGS += -I$(LIBUSB_PATH)/include -DIL_BIGENDIAN
 endif
 
 ifeq ($(TARGETENV), linuxmips)
 	TARGET_PREFIX := mipsel-linux-
 	OBJECTS += $(OBJDIR)usb_linux.o
-	CFLAGS += -I ../libusb
 	COMPILE = mipsel-linux-gcc 
 	CLEAN = clean_linux
-	LDFLAGS += -L$(LIBUSB_PATH)/.libs
-	LIBS= ../libusb/.libs/libusb.a
-	LIBS := -L$(LIBUSB_PATH)/.libs $(LIBUSB_PATH)/.libs/libusb.a
-	CFLAGS += -I$(LIBUSB_PATH)
+	LDFLAGS += -L$(LIBUSB_PATH)/lib
+	LIBS := -L$(LIBUSB_PATH)/lib $(LIBUSB_PATH)/lib/libusb.a
+	CFLAGS += -I$(LIBUSB_PATH)/include
 install: 
 	cp bcmdl $(SRCBASE)/router/mipsel-linux/target/bin
 endif
