diff -Bru kirin_rel_5_102_98_42_org/src/usbdev/usbdl/Makefile kirin_rel_5_102_98_42/src/usbdev/usbdl/Makefile
--- kirin_rel_5_102_98_42_org/src/usbdev/usbdl/Makefile	2012-06-21 17:38:00.000000000 -0400
+++ kirin_rel_5_102_98_42/src/usbdev/usbdl/Makefile	2012-09-14 14:52:18.440916904 -0400
@@ -40,20 +40,18 @@
 ifeq ($(TARGETENV), linuxmips_be) 
 	OBJECTS += $(OBJDIR)usb_linux.o
 	COMPILE = mips-linux-gcc 
-	LIBS := -L$(LIBUSB_PATH)/.libs $(LIBUSB_PATH)/.libs/libusb.a
-	CFLAGS += -I$(LIBUSB_PATH) -DIL_BIGENDIAN
+	LIBS := -L$(LIBUSB_PATH)/lib $(LIBUSB_PATH)/lib/libusb.a
+	CFLAGS += -I$(LIBUSB_PATH)/include -DIL_BIGENDIAN
 endif
 
 ifeq ($(TARGETENV), linuxmips)
 	TARGET_PREFIX := mipsel-linux-
 	OBJECTS += $(OBJDIR)usb_linux.o
-	CFLAGS += -I ../libusb
 	COMPILE = mipsel-linux-gcc 
 	CLEAN = clean_linux
-	LDFLAGS += -L$(LIBUSB_PATH)/.libs
-	LIBS= ../libusb/.libs/libusb.a
-	LIBS := -L$(LIBUSB_PATH)/.libs $(LIBUSB_PATH)/.libs/libusb.a
-	CFLAGS += -I$(LIBUSB_PATH)
+	LDFLAGS += -L$(LIBUSB_PATH)/lib
+	LIBS := -L$(LIBUSB_PATH)/lib $(LIBUSB_PATH)/lib/libusb.a
+	CFLAGS += -I$(LIBUSB_PATH)/include
 install: 
 	cp bcmdl $(SRCBASE)/router/mipsel-uclibc/target/bin
 endif
diff -Bru kirin_rel_5_102_98_42_org/src/wl/config/wlconfig_lx_wl_media kirin_rel_5_102_98_42/src/wl/config/wlconfig_lx_wl_media
--- kirin_rel_5_102_98_42_org/src/wl/config/wlconfig_lx_wl_media	2012-06-21 17:37:54.000000000 -0400
+++ kirin_rel_5_102_98_42/src/wl/config/wlconfig_lx_wl_media	2012-09-14 14:52:18.440916904 -0400
@@ -35,7 +35,8 @@
                  { USB_DEVICE(0x1eda, 0x0bdc) }, \
                  { USB_DEVICE(0x043E, 0x3001) }, \
                  { USB_DEVICE(0x04da, 0x23f2) }, \
-                 { USB_DEVICE(0x0471, 0x20E7) }
+                 { USB_DEVICE(0x0471, 0x20E7) }, \
+                 { USB_DEVICE(0x13b1, 0x003a) }
 
 
 ############################################
Only in kirin_rel_5_102_98_42/src/wl/config: wlconfig_lx_wl_media.orig
diff -Bru kirin_rel_5_102_98_42_org/src/wps/wpsapi/common/include/wps_api_osl.h kirin_rel_5_102_98_42/src/wps/wpsapi/common/include/wps_api_osl.h
--- kirin_rel_5_102_98_42_org/src/wps/wpsapi/common/include/wps_api_osl.h	2012-06-21 17:38:08.000000000 -0400
+++ kirin_rel_5_102_98_42/src/wps/wpsapi/common/include/wps_api_osl.h	2012-09-14 14:52:18.444918901 -0400
@@ -62,7 +62,7 @@
 extern char *wps_osl_get_short_adapter_name();
 extern int wps_osl_get_mac(uint8 *mac);
 extern char *wps_osl_get_ssid();
-extern uint32 wps_osl_init(void *cb_ctx, void *cb, const char *adapter_id);
+extern uint32 wps_osl_init(void *cb_ctx, void *cb, const char *adapter_id, const char* bridge_id);
 extern void wps_osl_deinit();
 extern void wps_osl_abort();
 extern uint32 wps_osl_setup_802_1x(uint8 *bssid);
diff -Bru kirin_rel_5_102_98_42_org/src/wps/wpsapi/common/include/wps_sdk.h kirin_rel_5_102_98_42/src/wps/wpsapi/common/include/wps_sdk.h
--- kirin_rel_5_102_98_42_org/src/wps/wpsapi/common/include/wps_sdk.h	2012-06-21 17:38:08.000000000 -0400
+++ kirin_rel_5_102_98_42/src/wps/wpsapi/common/include/wps_sdk.h	2012-09-14 14:52:18.444918901 -0400
@@ -277,6 +277,8 @@
  * Function : wps_api_open
  * Parameters :
  *		adapter_id - Adapter identifier
+ *		bridge_id Bridge adaptor name that the interface is connected too (the interface to
+ *		listen for events on)
  *		cb_ctx  - Optional. If provided, the WPS status callback
  *			will be called with this context.
  *		callback - WPS status callback.
@@ -303,10 +305,10 @@
  *			serial Number = "5678"
  */
 #ifdef WFA_WPS_20_TESTBED
-BCM_WPSAPI bool wps_api_open(const char *adapter_id, void *cb_ctx, fnWpsProcessCB callback,
+BCM_WPSAPI bool wps_api_open(const char *adapter_id, const char* bridge_id, void *cb_ctx, fnWpsProcessCB callback,
 	wps_devinf *devinf, wps20_testbed_inf *wps20_tbinf, bool ap_pin, bool version2);
 #else
-BCM_WPSAPI bool wps_api_open(const char *adapter_id, void *cb_ctx, fnWpsProcessCB callback,
+BCM_WPSAPI bool wps_api_open(const char *adapter_id, const char* bridge_id, void *cb_ctx, fnWpsProcessCB callback,
 	wps_devinf *devinf, bool ap_pin, bool version2);
 #endif
 
diff -Bru kirin_rel_5_102_98_42_org/src/wps/wpsapi/common/wps_api.c kirin_rel_5_102_98_42/src/wps/wpsapi/common/wps_api.c
--- kirin_rel_5_102_98_42_org/src/wps/wpsapi/common/wps_api.c	2012-06-21 17:38:07.000000000 -0400
+++ kirin_rel_5_102_98_42/src/wps/wpsapi/common/wps_api.c	2012-09-14 14:52:18.448920899 -0400
@@ -734,10 +734,10 @@
 /* wps_api_open function must be called first, before any other wps api call */
 BCM_WPSAPI bool
 #ifdef WFA_WPS_20_TESTBED
-wps_api_open(const char *adapter_id, void *cb_ctx, fnWpsProcessCB callback, wps_devinf *devinf,
+wps_api_open(const char *adapter_id, const char* bridge_id, void *cb_ctx, fnWpsProcessCB callback, wps_devinf *devinf,
 	wps20_testbed_inf *wps20_tbinf, bool ap_pin, bool b_v2)
 #else
-wps_api_open(const char *adapter_id, void *cb_ctx, fnWpsProcessCB callback, wps_devinf *devinf,
+wps_api_open(const char *adapter_id, const char* bridge_id, void *cb_ctx, fnWpsProcessCB callback, wps_devinf *devinf,
 	bool ap_pin, bool b_v2)
 #endif /* WFA_WPS_20_TESTBED */
 {
@@ -796,7 +796,7 @@
 	wps_api_status_cb(&wps_api_wksp->cb, wps_api_wksp->cb_ctx, WPS_STATUS_INIT, NULL);
 
 	/* WPS hook init for adapter and led (HW) */
-	if (wps_hook_init(cb_ctx, callback, adapter_id) == false) {
+	if (wps_hook_init(cb_ctx, callback, adapter_id, bridge_id) == false) {
 		TUTRACE((TUTRACE_ERR, "wps_api_open : Failed to initial wireless adapter.\n"));
 		return false;
 	}
diff -Bru kirin_rel_5_102_98_42_org/src/wps/wpsapi/common/wps_api_priv.h kirin_rel_5_102_98_42/src/wps/wpsapi/common/wps_api_priv.h
--- kirin_rel_5_102_98_42_org/src/wps/wpsapi/common/wps_api_priv.h	2012-06-21 17:38:07.000000000 -0400
+++ kirin_rel_5_102_98_42/src/wps/wpsapi/common/wps_api_priv.h	2012-09-14 14:52:18.448920899 -0400
@@ -35,7 +35,7 @@
 
 extern bool wps_hook_create_profile(const struct _wps_credentials *credentials);
 extern int wps_hook_get_mac(uint8 *mac);
-extern bool wps_hook_init(void *cb_ctx, void *cb, const char *adapter_id);
+extern bool wps_hook_init(void *cb_ctx, void *cb, const char *adapter_id, const char* bridge_id);
 extern void wps_hook_deinit();
 extern void wps_hook_abort();
 extern uint32 wps_hook_setup_802_1x(char *bssid);
diff -Bru kirin_rel_5_102_98_42_org/src/wps/wpsapi/common/wps_hooks.c kirin_rel_5_102_98_42/src/wps/wpsapi/common/wps_hooks.c
--- kirin_rel_5_102_98_42_org/src/wps/wpsapi/common/wps_hooks.c	2012-06-21 17:38:07.000000000 -0400
+++ kirin_rel_5_102_98_42/src/wps/wpsapi/common/wps_hooks.c	2012-09-14 14:52:18.448920899 -0400
@@ -61,9 +61,9 @@
 
 /* Initial HW related */
 bool
-wps_hook_init(void *cb_ctx, void *cb, const char *adapter_id)
+wps_hook_init(void *cb_ctx, void *cb, const char *adapter_id, const char * bridge_id)
 {
-	uint32 retVal = wps_osl_init(cb_ctx, cb, adapter_id);
+	uint32 retVal = wps_osl_init(cb_ctx, cb, adapter_id, bridge_id);
 	return ((retVal == WPS_OSL_SUCCESS) ? true : false);
 }
 
Only in kirin_rel_5_102_98_42/src/wps/wpsapi/linux: Makefile.orig
diff -Bru kirin_rel_5_102_98_42_org/src/wps/wpsapi/linux/wps_api_tester.c kirin_rel_5_102_98_42/src/wps/wpsapi/linux/wps_api_tester.c
--- kirin_rel_5_102_98_42_org/src/wps/wpsapi/linux/wps_api_tester.c	2012-06-21 17:38:08.000000000 -0400
+++ kirin_rel_5_102_98_42/src/wps/wpsapi/linux/wps_api_tester.c	2012-09-14 14:53:04.744059082 -0400
@@ -343,11 +343,11 @@
 
 	/* 2. Open WPS */
 #ifdef WFA_WPS_20_TESTBED
-	bRet = wps_api_open(STRP(wps->ifname), NULL, _wps_join_callback,
+	bRet = wps_api_open(STRP(wps->ifname), STRP(wps->brname), NULL, _wps_join_callback,
 		_wps_my_devinf(&my_devinf, wps->transport_uuid), &wps->wps20_tbinf, wps->b_appin,
 		wps->b_v2);
 #else
-	bRet = wps_api_open(STRP(wps->ifname), NULL, _wps_join_callback,
+	bRet = wps_api_open(STRP(wps->ifname), STRP(wps->brname), NULL, _wps_join_callback,
 		_wps_my_devinf(&my_devinf, wps->transport_uuid), wps->b_appin, wps->b_v2);
 #endif
 	if (bRet == FALSE) {
diff -Bru kirin_rel_5_102_98_42_org/src/wps/wpsapi/linux/wps_api_tester.h kirin_rel_5_102_98_42/src/wps/wpsapi/linux/wps_api_tester.h
--- kirin_rel_5_102_98_42_org/src/wps/wpsapi/linux/wps_api_tester.h	2012-06-21 17:38:08.000000000 -0400
+++ kirin_rel_5_102_98_42/src/wps/wpsapi/linux/wps_api_tester.h	2012-09-14 14:53:04.744059082 -0400
@@ -29,7 +29,8 @@
 
 	char bssid[6];
 	char ssid[33];
-	char ifname[16];
+    char ifname[16];
+    char brname[16];
 	char def_dhclient_pf[256];
 	char ip_addr[16];
 	char dhcp_cmd[256];
diff -Bru kirin_rel_5_102_98_42_org/src/wps/wpsapi/linux/wps_api_ui.c kirin_rel_5_102_98_42/src/wps/wpsapi/linux/wps_api_ui.c
--- kirin_rel_5_102_98_42_org/src/wps/wpsapi/linux/wps_api_ui.c	2012-06-21 17:38:08.000000000 -0400
+++ kirin_rel_5_102_98_42/src/wps/wpsapi/linux/wps_api_ui.c	2012-09-14 14:53:04.748061081 -0400
@@ -968,11 +968,16 @@
 			wps->b_ssid = TRUE;
 			WPS_PRINT(("SSID : %s", wps->ssid));
 		}
-		else if (!strcmp(cmd, "-if")) {
-			WPS_ARGC_CHECK();
-			val = argv[index++]; argc--;
-			wps_strncpy(wps->ifname, val, sizeof(wps->ifname));
-		}
+        else if (!strcmp(cmd, "-if")) {
+            WPS_ARGC_CHECK();
+            val = argv[index++]; argc--;
+            wps_strncpy(wps->ifname, val, sizeof(wps->ifname));
+        }
+        else if (!strcmp(cmd, "-br")) {
+            WPS_ARGC_CHECK();
+            val = argv[index++]; argc--;
+            wps_strncpy(wps->brname, val, sizeof(wps->brname));
+        }
 		else if (!strcmp(cmd, "-bssid")) {
 			WPS_ARGC_CHECK();
 			/* 
diff -Bru kirin_rel_5_102_98_42_org/src/wps/wpsapi/linux/wps_linux_osl.c kirin_rel_5_102_98_42/src/wps/wpsapi/linux/wps_linux_osl.c
--- kirin_rel_5_102_98_42_org/src/wps/wpsapi/linux/wps_linux_osl.c	2012-06-21 17:38:08.000000000 -0400
+++ kirin_rel_5_102_98_42/src/wps/wpsapi/linux/wps_linux_osl.c	2012-09-14 14:52:18.452922897 -0400
@@ -89,6 +89,8 @@
 WPS_OSL_T *wps_osl_wksp = NULL;
 
 static char ifname_lx[IFNAMSIZ] = "";
+static char ifname_br[IFNAMSIZ] = "";
+
 static uint8 peer_mac[6] = {0};
 static int eap_fd = -1; /* descriptor to raw socket  */
 static int ifindex = -1; /* interface index */
@@ -154,6 +156,12 @@
 
 	ifr.ifr_name[0] = '\0';
 
+	if (ifname_br[0] != '\0')
+    {
+	    strncpy(ifname, ifname_br, ifname_len);
+	    return 0;
+    }
+
 	if (!(fp = fopen(proc_net_dev, "r")))
 		return ret;
 
@@ -647,7 +655,15 @@
 		memcpy(peer_mac, bssid, 6);
 
 	memset(&ifr, 0, sizeof(ifr));
+
+	if (ifname_br[0])
+	{
+	    wps_strncpy(ifr.ifr_name, ifname_br, sizeof(ifr.ifr_name));
+	}
+	else
+	{
 	wps_strncpy(ifr.ifr_name, ifname_lx, sizeof(ifr.ifr_name));
+	}
 
 	err = ioctl(eap_fd, SIOCGIFINDEX, &ifr);
 	if (err < 0) {
@@ -857,7 +873,7 @@
 }
 
 uint32
-wps_osl_init(void *cb_ctx, void *cb, const char *adapter_id)
+wps_osl_init(void *cb_ctx, void *cb, const char *adapter_id, const char* bridge_id)
 {
 	char *ifname = (char*) adapter_id;
 	char dyn_name[IFNAMSIZ];
@@ -886,6 +902,11 @@
 		ifname = dyn_name;
 	}
 
+	if (bridge_id != NULL)
+	{
+	    wps_strncpy(ifname_br, bridge_id, sizeof(ifname_br));
+	}
+
 	/* Save this specific interface for further use */
 	wps_strncpy(ifname_lx, ifname, sizeof(ifname_lx));
 
@@ -942,6 +963,8 @@
 		free(wps_osl_wksp);
 		wps_osl_wksp = NULL;
 	}
+	memset(ifname_br, 0, sizeof(ifname_br));
+	memset(ifname_lx, 0, sizeof(ifname_lx));
 }
 
 void
diff -Bru kirin_rel_5_102_98_42_org/src/wps/wpscli/common/wpscli_wl.c kirin_rel_5_102_98_42/src/wps/wpscli/common/wpscli_wl.c
--- kirin_rel_5_102_98_42_org/src/wps/wpscli/common/wpscli_wl.c	2012-06-21 17:38:08.000000000 -0400
+++ kirin_rel_5_102_98_42/src/wps/wpscli/common/wpscli_wl.c	2012-09-14 14:52:18.456924897 -0400
@@ -739,7 +739,7 @@
 		proberesp->primDeviceOui = devinfo->primDeviceOui;
 		proberesp->primDeviceSubCategory = devinfo->primDeviceSubCategory;
 		strcpy(proberesp->deviceName, devinfo->deviceName);
-		proberesp->configMethods = devinfo->configMethods & ~WPS_CONFMET_PBC;
+		proberesp->configMethods = devinfo->configMethods;
 		proberesp->rfBand = devinfo->rfBand;
 		if (b_wps_version2) {
 			proberesp->version2 = devinfo->version2;
